### PR TITLE
QA: System property to override distribution

### PR DESCRIPTION
**Pull request by nik9000**
This configures all `qa` projects to use the distribution contained in
the `tests.distribution` system property if it is set. The goal is to
create a simple way to run tests against the default distribution which
has x-pack basic features enabled while not forcing these tests on all
contributors. You run these tests by doing something like:

```
./gradlew -p qa -Dtests.distribution=zip check
```

or

```
./gradlew -p qa -Dtests.distribution=zip bwcTest
```

x-pack basic *shouldn't* get in the way of any of these tests but
nothing is ever perfect so this also adds support for skipping tests
based on the distribution. For the yaml tests this looks like:

```
    - skip:
        distribution: zip
        reason: x-pack includes templates which breaks the output here
```

For java tests it looks something like:

```
assumeFalse("zip".equals(System.getProperty("tests.distribution")),
        "some reason");
```

Gradle makes sure to set the `tests.distribution` environment variable
when running tests whether or not you override the distribution on the
command line. This allows us to remove some `tests.rest.blacklist`
settings in `x-pack/qa` because those tests skip themselves based on the
distribution. I think this sort of thing substantially reduces the
"surprise factor" while working on x-pack.

### CODE REVIEW

File: qa/build.gradle [hunk 1]
```diff
@@ -4,7 +4,7 @@ import org.elasticsearch.gradle.test.RestIntegTestTask
 subprojects { Project subproj ->
   subproj.tasks.withType(RestIntegTestTask) {
     subproj.extensions.configure("${it.name}Cluster") { cluster ->
-      cluster.distribution = 'oss-zip'
+      cluster.distribution = System.getProperty('tests.distribution', 'oss-zip')
```
Comment (nik9000) [inline qa/build.gradle]: At first I thought this PR would be **just** this line. If only things were that simple....

File: test/framework/src/main/java/org/elasticsearch/test/junit/listeners/ReproduceInfoPrinter.java [hunk 2]
```diff
@@ -149,6 +149,7 @@ public ReproduceErrorMessageBuilder appendESProperties() {
             }
             appendOpt("tests.locale", Locale.getDefault().toLanguageTag());
             appendOpt("tests.timezone", TimeZone.getDefault().getID());
+            appendOpt("tests.distribution", System.getProperty("tests.distribution"));
```
Comment (nik9000) [inline test/framework/src/main/java/org/elasticsearch/test/junit/listeners/ReproduceInfoPrinter.java]: If you override the distribution on the command line then it is important to override it again. If you *don't* override it on the command line then setting this environment variable does nothing which is fine.

File: x-pack/qa/core-rest-tests-with-security/build.gradle [hunk 3]
```diff
@@ -15,9 +15,6 @@ integTestRunner {
     ['cat.aliases/10_basic/Empty cluster',
      'index/10_with_id/Index with ID',
      'indices.get_alias/10_basic/Get alias against closed indices',
-     'cat.templates/10_basic/No templates',
```
Comment (nik9000) [inline x-pack/qa/core-rest-tests-with-security/build.gradle]: These are skipped by the `skip` section in the test now which I find *much* easier to read.

**Comment by nik9000** [comment]
This is a WIP because I'm still experimenting with it locally.

**Comment by nik9000** [comment]
I think I *could* drop the `skip` section changes in favor of more blacklists but I prefer I think they make the tests more readable. If we feel like it is too big to backport to 6.3 then I'm happy to go the blacklist route and do the skip in a followup.

**Comment by nik9000** [comment]
> I think I could drop the skip section changes in favor of more blacklists but I prefer I think they make the tests more readable. If we feel like it is too big to backport to 6.3 then I'm happy to go the blacklist route and do the skip in a followup.

On second thought I think I will drop the `skip` enhancement. Those yml files are shared between Elasticsearch and its language clients are they aren't likely to like the change.

File: qa/build.gradle (hunk 1 above)
Comment (nik9000) [reply qa/build.gradle]: And now that I've reverted most of the rest, it is almost the only line!

File: qa/mixed-cluster/build.gradle [hunk 4]
```diff
@@ -57,6 +57,11 @@ for (Version version : bwcVersions.wireCompatible) {
   /* To support taking index snapshots, we have to set path.repo setting */
   tasks.getByName("${baseName}#mixedClusterTestRunner").configure {
     systemProperty 'tests.path.repo', new File(buildDir, "cluster/shared/repo")
+    systemProperty 'tests.rest.blacklist', [
```
Comment (nik9000) [inline qa/mixed-cluster/build.gradle]: No, no, this isn't right.

**Comment by nik9000** [comment]
I'm still working through getting a clean run of `./gradlew -p qa check -Dtests.distribution=zip`. Most things are working but it not everything. So far the only failing things are to do with the tests, not Elasticsearch.

**Comment by nik9000** [comment]
OK! I've got `./gradlew -p qa check -Dtests.distribution=zip` passing locally. I'm running `./gradlew -p qa bwcTest -Dtests.distribution=zip` to make sure that that passes as well but I suspect it will.

**Comment by nik9000** [comment]
OK! I got a successful run of `./gradlew -p qa bwcTest -Dtests.distribution=zip`!

**Review by rjernst — APPROVED** [review]
LGTM

File: qa/smoke-test-rank-eval-with-mustache/build.gradle [hunk 5]
```diff
@@ -26,3 +26,10 @@ dependencies {
     testCompile project(path: ':modules:lang-mustache', configuration: 'runtime')
 }
 
+/*
+ * One of the integration tests doesn't work with the zip distribution
```
[the remark this answers is not in the document]
Comment (nik9000) [reply qa/smoke-test-rank-eval-with-mustache/build.gradle]: Absolutely.

**Comment by bleskes** [comment]
It looks like this was backported. I removed the label.

**Pull request merged**